Streamline input implementation

mucaho opened the pull request:
EDIT:
**Current impact: **
* fixes _"Someone could hold down a key while we've lost focus, and then release once it's back, so we probably don't want to fire keyup events in such a case."_ part of #552 
* fixes all points of #1013 
* tries to fix #1043 , needs testing from a friendly soul with access to that :)
* **API change** of `findPointerTargetEventByComponent`, and all documented for clarity, but only internally used `dispatch` methods

mucaho committed:
Streamline input implementation

* **API change:** Changes the signature and functionality of
Crafty.findPointerEventTargetByComponent:
This method now only finds the entity. A newly introduced
Crafty.augmentPointerEvent augments pointer events with
coordinates from real world space.
* Possibly fixes Click on touch devices (#1043):
Trigger Click event in touchDispatch if distance from start
to end of touch is smaller than some threshold.
* Augment MouseScroll event with real coordinates.
* Extract common functionality (e.g. preventDefault) into a
EventDispatcher prototype.
* Introduce lockFocus to EventDispatchers (Mouse & Keyboard),
so that elements can force to receive all subsequent input events.
Make viewport.mouselook and Draggable use this. This fixes one
part of issue #1013.
* Route all mouse events without target to the global "Mouse"
system, which controls and viewport can bind to. This fixes
the other part of issue #1013.
* Properly handle stage focus loss:
Add resetMouseDown which similary to resetKeyDown triggers ButtonUp
events for currently pressed buttons to currently locked target.
ResetKeyDown and resetMouseDown are called on 3 occasions -
on Crafty.pause, on Crafty.stop and on stage focus loss.
This fixes invalid drag state of Draggable entities and mouselooked
viewport when user clicks outside stage while dragging (related
to #1013).

starwed commented:
>Route all mouse events without target to the global "Mouse" system,
which controls and viewport can bind to.

I wonder if it would make sense to route *all* events to such a system, with information about whether there was a target or not.

Aside: if this ends up changing the api in a backwards-incompatible way, let's make sure to prep another point release before merging.

mucaho commented:
> I wonder if it would make sense to route *all* events to such a system, with information about whether there was a target or not.

Sure, but what would be the benefits?

> Aside: if this ends up changing the api in a backwards-incompatible way, let's make sure to prep another point release before merging.

EDIT: Initially I thought there wouldn't be many breaking changes, but alas, there are :)

mucaho committed:
Add Mouse and KeyboardStates

Add MouseState to Mouse component, Mouse system and system's
focused element.
Add KeyboardState to Keyboard component, Keyboard system and
the system's focused element.

This State functionality records MouseButtonDowns & MouseButtonUps
or KeyboardDowns & KeyboardUps respectively.
The State functionality offers triggerKey / triggerMouseButton
methods which propagate the incoming event to the default trigger
method, if the action is valid (e.g. don't propagate KeyUp if KeyDown
wasn't received beforehand).

That means that the methods Crafty.resetKeyDown and
Crafty.resetMouseDown can be called as often as needed (e.g. due to
loss of window focus) without any unwanted side effects.

The Systems' focused target system is wrapped into such a State
wrapper, since we want to preserve correct key / button state for it
too. (We can't expect a target system, like viewport, to handle
correct key / mouse state itself).
This wrapping is not needed for focused target entities, since e.g.
Draggable transitively requires MouseState.

This commit fixes part of #552, quote: "Someone could hold down a key
while we've lost focus, and then release once it's back, so we probably
don't want to fire keyup events in such a case."

Commit adds tests and complex, "real world" examples for locking.

mucaho commented:
PR is ready, so starwed if you'd like to take a look at it, it would be greatly appreciated! If you are going to do so, you could omit whitespace with `?w=1` in the diff url. Commit messages should describe in detail all changes. See [current impact](https://github.com/craftyjs/Crafty/pull/1116#issue-221991996), to summarize it's a small API change that fixes some open issues. What do you think, would it make sense to include these fixes in next minor release, too?

What's planned in next PR, will modify API and semantics considerably.
* Split everything into appropriate files. Dispatchers into own file. Mouse & Keyboard system and components into own file. Event setup into own file.
* Let `Crafty.keydown`, `Crafty.resetKeyDown`, `Crafty.lastEvent` live under the appropriate dispatcher namespace. So the whole logic is condensed in each dispatcher.
* Introduce `lastKeyEvent` to KeyboardState and repurpose `Crafty.lastEvent` as `lastMouseEvent` into MouseState
* Additionally, convert the dispatcher itself to a system and let it inherit from MouseState, KeyState. The dispatcher will trigger all events on itself, no matter where it routes them afterwards. This will allow users to always be notified on newest input events.
* Motivation: This "layered" architecture is needed, because we can have the following situation:
  * A KeyDown has been dispatched globally. We can't know which entity modified its state based on this.
  * Another entity requires the lock.
  * We now need to dispatch a KeyUp event globally, since the initial entity won't receive a KeyUp event until the other entity released the lock. It would be stuck in an inconsistent state.
* Trigger Keyboard events only on entities with the Keyboard component and on the Keyboard system. (depends on #1117). Since we need to dispatch artificial KeyUp events (see previous point about motivation), we have to make sure not to cause inconsistent state. Thus only entities with KeyboardState component will receive it, in order to shield them from state corruption.
* Suggestions are welcome!

After that PR, figure out how to make the control system and components get their input from the entity-specific keyboard and mouse states.

starwed commented:
>Trigger Keyboard events only on entities with the Keyboard component and on the Keyboard system. (depends on #1117). Since we need to dispatch artificial KeyUp events (see previous point about motivation), we have to make sure not to cause inconsistent state. Thus only entities with KeyboardState component will receive it, in order to shield them from state corruption.

The main motivation of the control system I added was to move away from entities listening directly to keyboard/mouse/etc events.

starwed commented:
Oops, accidentally hit close!  

>Trigger Keyboard events only on entities with the Keyboard component and on the Keyboard system. (depends on #1117). Since we need to dispatch artificial KeyUp events (see previous point about motivation), we have to make sure not to cause inconsistent state. Thus only entities with KeyboardState component will receive it, in order to shield them from state corruption.

The main motivation of the control system I added was to move away from entities listening directly to keyboard/mouse/etc events.  A big part of that was avoiding things like artificial KeyUp events; you'd have the control system dispatch "TriggerUp" events (etc) instead.

mucaho commented:
> The main motivation of the control system I added was to move away from entities listening directly to keyboard/mouse/etc events. A big part of that was avoiding things like artificial KeyUp events; you'd have the control system dispatch "TriggerUp" events (etc) instead.

Right, but that only takes care of control components. (Although "only" in this case covers probably to the majority of use-cases). Users may want to listen to "KeyDown" and "KeyUp" events for their own implementation, with this proposal it would be safe to do so.

starwed commented:
I'd argue that the concept of locking events to a specific entity should happen in the control layer not the input layer.  That way you *can* opt in to getting the raw events, which will be as close a representation to the actual input as we can get.

Of course, we don't actually run pointer events through the control system right now...  :)

If we were to move towards that, the rough idea is this sort of structure:
- The input layer handles normalizing how browsers pass in the key/pointer/etc events.  This is pretty much all the existing [keyboardDispatch](https://github.com/craftyjs/Crafty/blob/develop/src/controls/inputs.js#L634-L674) method does right now, but mouseDispatch is doing a bit more.  In my idealized version, it just produces a single event that passes through the crafty stage coordinates, normalized button information, and the original dom target.  Any logic should be happening at a later stage.
- The control layer would then handle interpreting the normalized data and conveying it to entities.  For pointer events, I think that would include finding the target of the event, which is part of dispatch right now, as well as any notion of locking.

An advantage of this approach is that it should be easier to add support for multiple types of input mapping to the same control behavior.  (For instance, extending the current viewport mouselook to be controllable from the keyboard.)

I'm not 100% convinced that my way is right, (and there could be big obstacles I'm not seeing, e.g. adding the notion of locking to the current control system) but I *am* certain we shouldn't work in opposite directions on this.  :)  I'd feel more comfortable if we hashed out the preferred approach before adding any additional public apis.  What are your thoughts?

mucaho commented:
Thanks for taking the time to think this through!

----------

> I'd argue that the concept of locking events to a specific entity should happen in the control layer not the input layer. That way you can opt in to getting the raw events, which will be as close a representation to the actual input as we can get.

This case is handled by 

> The dispatcher will trigger all events on itself, no matter where it routes them afterwards _[or if something is locked currently]_. This will allow users to always be notified on newest _[raw]_ input events.

-----------------------------------------------------

> An advantage of this approach is that it should be easier to add support for multiple types of input mapping to the same control behavior. (For instance, extending the current viewport mouselook to be controllable from the keyboard.)

Seems like a good idea! Adding an additional layer of indirection on the input events shouldn't conflict with this. On the contrary, it should serve as a foundation for the control system. I firmly believe that the input system itself should filter out invalid events, even if it only serves to let users have some other safe implementation outside the control system.

-------------------------------------------

> The control layer would then handle _[conveying data]_ to entities. For pointer events, I think that would include finding the target of the event, which is part of dispatch right now, as well as any notion of locking.

Now this point is interesting, and I agree that routing and locking of events could be handled by the logic of the control system. However, for this to work the rest of current control components (Draggable and mouselook) need to be adapted to use the control system.

-----------------------------------------------

Are you content to go forward with the following approach?
* Dispatch all mouse events only to the Mouse system and Mouse components
* Dispatch all key events only to the Keyboard system and Keyboard components.
* These systems and components would trigger only valid input events (due to the validation of key / mouse state)
* Control system would either listen to the appropriate system or component and then do its routing based on its control logic

If you are ok with this, i would remove the locking and routing behavior from the input system, but keep the rest of this PR intact.

mucaho commented:
Thought about it a little bit more: if routing and locking is shifted to the controls there is no need for key state validation on the level of individual components. It suffices if Keyboard dispatch takes care of that on the global level and if it calls resetKeyDown when needed.
There may be some fixes that could be extracted from here, but other than it's obsolete
